File: go/tests/logistics_board_test.go
// logistics_board_test.go
//
// Verifies the CircleAI.Logistics port (logistics_board.go): shipment/vehicle
// register (blank-id error), GetShipment, Vehicles id-ordering, and PlanRoute
// (distance sum, cost = totalKm*CostPerKm as Decimal, sequential plan ids,
// blank/unknown-vehicle errors).

package circleai_test

import (
	"testing"
	"time"

	circleai "github.com/bhengubv/CircleAI/go"
)

func TestLogistics_RegisterAndVehicles(t *testing.T) {
	b := circleai.NewInMemoryLogisticsBoard()
	pickup := time.Date(2026, 7, 1, 6, 0, 0, 0, time.UTC)
	if err := b.RegisterShipment(circleai.Shipment{ShipmentId: "sh1", Origin: "JHB", Destination: "CPT", WeightKg: 500, VolumeM3: 2, Incoterm: "DAP", PickupAtUtc: pickup}); err != nil {
		t.Fatalf("register shipment: %v", err)
	}
	if err := b.RegisterShipment(circleai.Shipment{ShipmentId: " "}); err == nil {
		t.Fatalf("blank ShipmentId must error")
	}
	if err := b.RegisterVehicle(circleai.Vehicle{VehicleId: "v2", CapacityKg: 1000, CapacityM3: 10, CostPerKm: 5}); err != nil {
		t.Fatalf("register vehicle: %v", err)
	}
	if err := b.RegisterVehicle(circleai.Vehicle{VehicleId: "v1", CapacityKg: 2000, CapacityM3: 20, CostPerKm: 8}); err != nil {
		t.Fatalf("register vehicle: %v", err)
	}
	if err := b.RegisterVehicle(circleai.Vehicle{VehicleId: ""}); err == nil {
		t.Fatalf("blank VehicleId must error")
	}

	if s, ok := b.GetShipment("sh1"); !ok || s.Destination != "CPT" {
		t.Fatalf("get shipment = %+v ok=%v", s, ok)
	}
	vs := b.Vehicles()
	if len(vs) != 2 || vs[0].VehicleId != "v1" || vs[1].VehicleId != "v2" {
		t.Fatalf("vehicles id order wrong: %+v", vs)
	}
}

func TestLogistics_PlanRoute(t *testing.T) {
	b := circleai.NewInMemoryLogisticsBoard()
	_ = b.RegisterVehicle(circleai.Vehicle{VehicleId: "v1", CapacityKg: 1000, CapacityM3: 10, CostPerKm: 2.5})

	legs := []circleai.RouteLeg{
		{FromCode: "JHB", ToCode: "BFN", DistanceKm: 400},
		{FromCode: "BFN", ToCode: "CPT", DistanceKm: 600},
	}
	plan, err := b.PlanRoute("v1", legs)
	if err != nil {
		t.Fatalf("plan route: %v", err)
	}
	if plan.TotalDistanceKm != 1000 {
		t.Fatalf("total km = %v, want 1000", plan.TotalDistanceKm)
	}
	// 1000 km * 2.5 = 2500.
	if !plan.EstimatedCost.Equal(circleai.DecimalFromInt(2500)) {
		t.Fatalf("cost = %s, want 2500", plan.EstimatedCost)
	}
	if len(plan.Legs) != 2 {
		t.Fatalf("plan legs = %d, want 2", len(plan.Legs))
	}
	// Second plan gets a distinct sequential id.
	plan2, _ := b.PlanRoute("v1", legs)
	if plan2.PlanId == plan.PlanId {
		t.Fatalf("plan ids should be distinct: %q == %q", plan.PlanId, plan2.PlanId)
	}

	if _, err := b.PlanRoute(" ", legs); err == nil {
		t.Fatalf("blank vehicleId must error")
	}
	if _, err := b.PlanRoute("ghost", legs); err == nil {
		t.Fatalf("unknown vehicle must error")
	}
}

File: typescript/tests/logistics_board.test.ts
// logistics_board.test.ts
// Verifies the CircleAI.Logistics port: shipment + vehicle registration
// (id-ordered vehicles) and the distance × cost-per-km route planner.

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import {
  InMemoryLogisticsBoard,
  LogisticsDomainContext,
  shipment,
  vehicle,
  routeLeg,
} from "../src/logistics/index";

describe("InMemoryLogisticsBoard", () => {
  it("registers shipments and vehicles (vehicles ordered by id)", () => {
    const b = new InMemoryLogisticsBoard();
    b.registerShipment(shipment("s1", "JNB", "CPT", 1000, 5, "EXW", new Date("2026-05-01T00:00:00Z")));
    b.registerVehicle(vehicle("vB", 5000, 30, 12));
    b.registerVehicle(vehicle("vA", 3000, 20, 9));
    assert.equal(b.getShipment("s1")?.destination, "CPT");
    assert.deepEqual(
      b.vehicles.map((v) => v.vehicleId),
      ["vA", "vB"],
    );
    assert.throws(() => b.registerShipment(shipment("", "a", "b", 1, 1, "x", new Date())), /ShipmentId required/);
    assert.throws(() => b.registerVehicle(vehicle("", 1, 1, 1)), /VehicleId required/);
  });

  it("plans routes: total distance and cost, unknown vehicle throws", () => {
    const b = new InMemoryLogisticsBoard();
    b.registerVehicle(vehicle("v1", 3000, 20, 10));
    const plan = b.planRoute("v1", [routeLeg("JNB", "BFN", 400), routeLeg("BFN", "CPT", 1000)]);
    assert.equal(plan.totalDistanceKm, 1400);
    assert.equal(plan.estimatedCost, 14000); // 1400 * 10
    assert.equal(plan.planId, "plan-1");
    assert.equal(plan.legs.length, 2);
    const plan2 = b.planRoute("v1", [routeLeg("A", "B", 100)]);
    assert.equal(plan2.planId, "plan-2");
    assert.throws(() => b.planRoute("ghost", []), /Unknown vehicle 'ghost'/);
    assert.throws(() => b.planRoute(" ", []), /vehicleId required/);
  });

  it("domain context exposes prompt + compliance + tools", () => {
    assert.ok(LogisticsDomainContext.systemPromptSnippet.includes("[DOMAIN: Logistics]"));
    assert.deepEqual(LogisticsDomainContext.complianceFlags, ["RTMS", "SARS_Customs", "AARTO", "POPIA", "Incoterms_2020"]);
    assert.deepEqual(LogisticsDomainContext.suggestedTools, ["route_planner", "fleet_tracker", "customs_portal", "analytics"]);
  });
});
